src/core/providers/ogr/qgsogrprovider.cpp

```diff
@@ -1117,16 +1135,56 @@ unsigned char *QgsOgrProvider::getGeometryPointer( OGRFeatureH fet )
 
 
 QgsRectangle QgsOgrProvider::extent() const
+{
+  return extent3D().toRectangle();
+}
+
+void QgsOgrProvider::compute3DExtentSlowly( QgsOgrLayer *ogrLayer, OGREnvelope3D *extent ) const
+{
+  gdal::ogr_feature_unique_ptr f;
+
+  ogrLayer->ResetReading();
+  while ( f.reset( ogrLayer->GetNextFeature() ), f )
+  {
+    OGRGeometryH g = OGR_F_GetGeometryRef( f.get() );
+    if ( g && !OGR_G_IsEmpty( g ) )
+    {
+      OGREnvelope3D env;
+      OGR_G_GetEnvelope3D( g, &env );
+
+      extent->MinX = std::min( extent->MinX, env.MinX );
+      extent->MinY = std::min( extent->MinY, env.MinY );
+      extent->MaxX = std::max( extent->MaxX, env.MaxX );
+      extent->MaxY = std::max( extent->MaxY, env.MaxY );
+      extent->MinZ = std::min( extent->MinZ, env.MinZ );
+      extent->MaxZ = std::max( extent->MaxZ, env.MaxZ );
```

Comment:
if there's a mix of 2D/3D objects in the layer, MinZ/MaxZ will be 0 on 2D objects

Reply:
Is there a reason to obtain 0 for non existent values? Do you think gdal can instead return nan?

May be to solve this, we can retrieve the 2D/3D state in each row?

Reply:
I will try using `OGR_G_Is3D`

Reply:
> Is there a reason to obtain 0 for non existent values? Do you think gdal can instead return nan?

I'm not sure of the reason besides being what is implemented... Note that NaN is also tricky to use with std::min() / std::max() (cf https://stackoverflow.com/questions/72296440/how-to-prevent-stdmin-and-max-to-return-nan-if-the-first-element-of-the-array)

> May be to solve this, we can retrieve the 2D/3D state in each row? I will try using OGR_G_Is3D

yes, that's what I had in mind